### PR TITLE
Policy authorization options for NV indexes

### PR DESCRIPTION
**Pull request by idesai**
Adding policy authorization options for NV Index read/ write/ delete operations

### CODE REVIEW

**Comment by idesai** [comment]
Added tests for this PR

File: tools/tpm2_nvdefine.c [hunk 1]
```diff
@@ -95,7 +102,25 @@ static int nv_space_define(tpm_nvdefine_ctx *ctx) {
...
+        public_info.t.nvPublic.attributes.TPMA_NV_POLICYWRITE = 1;
+    }
+    if (ctx->enforce_delete_policy) {
+        public_info.t.nvPublic.attributes.TPMA_NV_POLICY_DELETE = 1;
+    }
+
+    if (!ctx->policy_file_flag) {
+        public_info.t.nvPublic.authPolicy.t.size = 0;
+    } else {
+        public_info.t.nvPublic.authPolicy.t.size  = sizeof(public_info.t.nvPublic.authPolicy) - 2;
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c]: on line 60, do TPM2B_EMPTY INIT
```
 TPM2B_NV_PUBLIC public_info = TPM2B_EMPTY_INIT;
```
Then here, just do:
```
if (ctx->policy_file_flag) {
   public_info.t.nvPublic.authPolicy.t.size = BUFFER_SIZE(TPM2B_NV_PUBLIC, <buffername>);
    files_load_bytes_from_file();
   ...
}
```

This gets rid of the if clause of your branch.

File: tools/tpm2_nvdefine.c [hunk 2]
```diff
@@ -130,6 +155,10 @@ static bool init(int argc, char* argv[], tpm_nvdefine_ctx *ctx) {
         { "handlePasswd", required_argument, NULL, 'P' },
         { "indexPasswd" , required_argument, NULL, 'I' },
         { "passwdInHex" , no_argument,       NULL, 'X' },
+        { "policy-file" , required_argument, NULL, 'L' },
+        { "enforce-read-policy", no_argument, NULL,'r'},
+        { "enforce-write-policy", no_argument, NULL,'w'},
+        { "enforce-delete-policy", no_argument, NULL,'d'},
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c]: Line these up.

File: tools/tpm2_nvdefine.c [hunk 3]
```diff
@@ -95,7 +102,25 @@ static int nv_space_define(tpm_nvdefine_ctx *ctx) {
 
     // Now set the attributes.
     public_info.t.nvPublic.attributes.val = ctx->nvAttribute;
-    public_info.t.nvPublic.authPolicy.t.size = 0;
+    if (ctx->enforce_read_policy) {
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c]: No need for branches...here and the others below:
```
 public_info.t.nvPublic.attributes.TPMA_NV_POLICYREAD = ctx->enforce_read_policy;
```

File: test/system/test_tpm2_nv.sh [hunk 4]
```diff
@@ -85,3 +86,17 @@ if [ $? != 0 ];then
 else
  echo "release the nv index OK!"
 fi
+
+echo "f28230c080bbe417141199e36d18978228d8948fc10a6a24921b9eba6bb1d988" \
+| xxd -r -p > policy.bin
+tpm2_nvdefine -x 0x1500016 -a 0x40000001 -s 32 -t 0x2000A -L policy.bin -r -w
+tpm2_nvlist | grep 0x1500016 -A5 | grep Auth | grep -o ": [a-zA-Z0-9]\{1,\}" | \
+grep -o "[a-zA-Z0-9]\{1\}" | xxd -r -p >test.bin
+tpm2_nvrelease -x 0x1500016 -a 0x40000001
```
Comment (williamcroberts) [inline test/system/test_tpm2_nv.sh]: Should we be checking the return codes above? Should we let it fail all the way to here?

File: test/system/test_tpm2_nv.sh [hunk 5]
```diff
@@ -85,3 +86,17 @@ if [ $? != 0 ];then
...
+cmp test.bin policy.bin -s
+return_val=$?
+if [ $return_val != 0 ];then
+	echo "nvdefine with policy in authorization structure failed"
```
Comment (williamcroberts) [inline test/system/test_tpm2_nv.sh]: What's the convention here? I think its 2 spaces looking within this file.

Comment (williamcroberts) [inline test/system/test_tpm2_nv.sh]: Granted it very inconsistent.

File: test/system/test_tpm2_nv.sh [hunk 6]
```diff
@@ -85,3 +86,30 @@ if [ $? != 0 ];then
...
+  cmp test.bin policy.bin -s
+  if [ $? != 0 ];then
+   echo "nvdefine with policy in authorization structure failed"
+  fi
+ fi
+fi
+if [ $return_val == 0 ];then
+ tpm2_nvrelease -x 0x1500016 -a 0x40000001
+ return_val=$?
+fi
```
Comment (williamcroberts) [inline test/system/test_tpm2_nv.sh]: same if fail then cleanup and bail

**Pull request merged**